### PR TITLE
[SPARK-47176][SQL] Have a ResolveAllExpressionsUpWithPruning helper function

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR proposes to have a new helper function to help tree traversal: `resolveExpressionsUpWithPruning`. This helper function traverses all expressions of a query tree bottom up and skipping the subtree if the condition returns false.
### Why are the changes needed?

Without this helper function, a developer will need to combine `plan.resolveOperatorsUpWithPruning` and `case p: LogicalPlan => p.transformExpressionsUpWithPruning` to achieve the same thing.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UT

### Was this patch authored or co-authored using generative AI tooling?

NO

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala [hunk 1]
```diff
@@ -250,6 +250,29 @@ trait AnalysisHelper extends QueryPlan[LogicalPlan] { self: LogicalPlan =>
     }
   }
 
+  /**
+   * Recursively transforms the expressions of a tree, first to all of its
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala:254]: I think we should highlight that the traversal order is bottom-up for both logical plan treea and expression tree.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala]: Tried to rephrase the wording to emphasize the `bottom-up` over all the expressions of all the tree node.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**